---
title: "Single Cell Transcriptomics with Python"
---

## Teachers

- Alex Russell Lederer [ORCiD](https://orcid.org/0000-0001-6381-5088)
- Geert van Geest [ORCiD](https://orcid.org/0000-0002-1561-078X)
- Tania Wyss [ORCiD](https://orcid.org/0000-0003-2641-0895)

## Attribution

Parts of this course are inspired by the [Single Cell Best Practices Guide](https://www.sc-best-practices.org/preamble.html) and previous R courses from the SIB on [Single Cell Transcriptomics](https://sib-swiss.github.io/single-cell-training/) by Tania Wyss, Rachel Marcone-Jeitziner, Geert van Geert, and Patricia Palagi.

## Overview
Single-cell RNA sequencing (scRNA-seq) can measure the gene expression of complex biological systems at the level of individual cells, enabling scientists to generate detailed tissue atlases describing the transcriptomic profiles of thousands or even millions of cells. While scRNA-seq has become a popular technique in diverse fields of biological research, the required expertise for handling such datasets has restricted its use among the larger scientific community. The aim of this 3-day course is to empower researchers to start applying the fundamental scRNA-seq analysis pipeline to their own data. We will outline how to design and interpret results of a scRNA-seq dataset and explore the basics of preprocessing and analysis in Python on real data. We will discuss common concerns in the field, including preprocessing choices, dimensionality reduction, cell type clustering and identification, batch effect correction, pseudotime, and RNA velocity methods. The course will be taught in Python.

## License & copyright

**License:** [CC BY-SA 4.0](https://github.com/sib-swiss/single-cell-training/blob/master/LICENSE.md)

**Copyright:** [SIB Swiss Institute of Bioinformatics](https://www.sib.swiss/)

## Learning outcomes

### General learning outcomes

By the end of the course, participants will be able to: 

- Run Cell Ranger
- Evaluate the quality of a scRNA-seq experiment
- Perform scanpy analysis on their own data
- Confidently communicate about how to overcome potential bottlenecks

### Learning outcomes explained

To reach the general learning outcomes above, we have set a number of smaller learning outcomes. Each chapter starts with these smaller learning outcomes. Use these at the start of a chapter to get an idea what you will learn. Use them also at the end of a chapter to evaluate whether you have learned what you were expected to learn.

## Learning experiences

To reach the learning outcomes we will use lectures, exercises, polls, and group work. During exercises, you are free to discuss with other participants. During lectures, focus on the lecture only.

### Exercises

Each block has practical work involved. Some more than others. The practicals are subdivided into chapters, and we'll have a (short) discussion after each chapter. All answers to the practicals are incorporated, but they are hidden. Do the exercise first by yourself, before checking out the answer. If your answer is different from the answer in the practicals, try to figure out why they are different.