<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Cell type annotation and visualization – Single-cell transcriptomics with Python</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assets/SIB_logo.svg" rel="icon" type="image/svg+xml">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-4d7b8eec1c60cc33e64fe606aab9e7fe.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro">
<!-- Matomo -->
<script>
  var _paq = window._paq = window._paq || [];
  /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="https://matomo.sib.swiss/";
    _paq.push(['setTrackerUrl', u+'matomo.php']);
    _paq.push(['setSiteId', '220']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
  })();
</script>
<!-- End Matomo Code -->


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Single-cell transcriptomics with Python</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../precourse_preparations.html"> 
<span class="menu-text">Pre-course preparations</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../course_schedule.html"> 
<span class="menu-text">Course schedule</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">    
        <li>
    <a class="dropdown-item" href="../ipynb/day1-1_setup.html">
 <span class="dropdown-text">Setup</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day1-1_cellranger.html">
 <span class="dropdown-text">Cell Ranger</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day1-2_analysis_tools_qc.html">
 <span class="dropdown-text">Analysis tools and quality control (QC)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day1-3_normalization_scaling.html">
 <span class="dropdown-text">Normalization</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">    
        <li>
    <a class="dropdown-item" href="../ipynb/day2-1_dimensionality_reduction.html">
 <span class="dropdown-text">Dimensionality Reduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day2-2_integration.html">
 <span class="dropdown-text">Data Integration</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day2-3_clustering.html">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day2-4_visualization.html">
 <span class="dropdown-text">Cell type annotation and visualization</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">    
        <li>
    <a class="dropdown-item" href="../ipynb/day3-1_tf.html">
 <span class="dropdown-text">Gene regulatory analysis with pySCENIC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day3-3_trajectory_analysis.html">
 <span class="dropdown-text">Trajectory Inference and Pseudotime</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day3-4_velocity1.html">
 <span class="dropdown-text">RNA velocity with scvelo</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ipynb/day3-4_velocity2.html">
 <span class="dropdown-text">RNA velocity with VeloCycle</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sib-swiss/single-cell-python-training/"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Cell type annotation and visualization</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#automated-cell-type-annotation" id="toc-automated-cell-type-annotation" class="nav-link active" data-scroll-target="#automated-cell-type-annotation">Automated cell type annotation</a></li>
  <li><a href="#another-way-to-annotate-with-label-transfer-from-a-reference-dataset" id="toc-another-way-to-annotate-with-label-transfer-from-a-reference-dataset" class="nav-link" data-scroll-target="#another-way-to-annotate-with-label-transfer-from-a-reference-dataset">Another way to annotate: with label transfer from a reference dataset!</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Cell type annotation and visualization</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<style>
    .large-link {
        font-size: 30px;
        font-family: Arial, sans-serif;
        color: #333;
        text-decoration: none;
    }
</style>
<p><a href="../assets/pdf/09_cell_type_annotation_and_visualization_day2.pdf" target="_blank" class="large-link">Download Presentation: Cell type annotation and visualization</a></p>
<div id="da6a6f44-4a88-4bb4-aa70-843434d02d8b" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd <span class="co"># for handling data frames (i.e. data tables)</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np <span class="co"># for handling numbers, arrays, and matrices</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt <span class="co"># plotting package</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns <span class="co"># plotting package</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Exercise 0</strong>: Before we continue in this notebook with the next steps of the analysis, we need to load our results from the previous notebook using the <code>sc.read_h5ad</code> function and assign them to the variable name <code>adata</code>. Give it a try!</p>
<div id="6ded009a-3d69-46ff-bc62-51b059a25388" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"PBMC_analysis_SIB_tutorial5.h5ad"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cf9ca046-bb99-43a9-aa0d-37b509759d24" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>AnnData object with n_obs × n_vars = 5465 × 3000
    obs: 'sample', 'n_counts', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'is_doublet', 'S_score', 'G2M_score', 'phase', 'leiden', 'leiden_res1', 'leiden_res0_1', 'leiden_res0_5', 'leiden_res2'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'
    uns: 'hvg', 'leiden', 'leiden_colors', 'leiden_res0_1_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'leiden_res2_colors', 'log1p', 'neighbors', 'pca', 'phase_colors', 'sample_colors', 'umap'
    obsm: 'X_pca', 'X_pcahm', 'X_umap'
    varm: 'PCs'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<p>Before proceeding with marker gene analysis and cell type annotation, restore the raw version of the data, add the necessary annotations, and normalize the counts:</p>
<div id="31d1782a-b522-4ef7-aae8-6313a58df860" class="cell" data-executioninfo="{&quot;elapsed&quot;:371,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340449533,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="b94ec1f2-fe65-46be-a531-9daec9f4e464" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>adata_raw <span class="op">=</span> sc.read_h5ad(<span class="st">"PBMC_analysis_SIB_tutorial.h5ad"</span>) <span class="co"># raw data before selecting highly variable genes</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>shared_bcs <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(adata.obs.index) <span class="op">&amp;</span> <span class="bu">set</span>(adata_raw.obs.index))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>adata_raw <span class="op">=</span> adata_raw[shared_bcs].copy()</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[shared_bcs].copy()</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>adata_raw_norm <span class="op">=</span> adata_raw.copy()</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata_raw_norm, target_sum<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata_raw_norm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3b89a3ee-0f8b-4c68-b9c8-2d837eff5e41" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obs[<span class="st">"leiden"</span>] <span class="op">=</span> adata.obs[<span class="st">"leiden_res1"</span>]</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obsm[<span class="st">"X_pca"</span>] <span class="op">=</span> adata.obsm[<span class="st">"X_pca"</span>]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obsm[<span class="st">"X_pcahm"</span>] <span class="op">=</span> adata.obsm[<span class="st">"X_pcahm"</span>]</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obsm[<span class="st">"X_umap"</span>] <span class="op">=</span> adata.obsm[<span class="st">"X_umap"</span>]</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obsp[<span class="st">"connectivities"</span>] <span class="op">=</span> adata.obsp[<span class="st">"connectivities"</span>]</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obsp[<span class="st">"distances"</span>] <span class="op">=</span> adata.obsp[<span class="st">"distances"</span>]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.uns[<span class="st">"neighbors"</span>] <span class="op">=</span> adata.uns[<span class="st">"neighbors"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s use a simple method implemented by scanpy to find marker genes by the Leiden cluster.</p>
<div id="5dec4127-2d0d-419a-9335-9bd045129336" class="cell" data-executioninfo="{&quot;elapsed&quot;:15954,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340476850,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="a83e0455-6ecf-4363-d635-cd20d00a6549" data-execution_count="7">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    adata_raw_norm, use_raw<span class="op">=</span><span class="va">False</span>, groupby<span class="op">=</span><span class="st">"leiden"</span>, method<span class="op">=</span><span class="st">"wilcoxon"</span>, key_added<span class="op">=</span><span class="st">"dea_leiden"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'names'] = self.var_names[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:398: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'scores'] = scores[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:401: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals'] = pvals[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:411: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:422: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'logfoldchanges'] = np.log2(
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'names'] = self.var_names[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:398: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'scores'] = scores[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:401: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals'] = pvals[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:411: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:422: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'logfoldchanges'] = np.log2(
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'names'] = self.var_names[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:398: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'scores'] = scores[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:401: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals'] = pvals[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:411: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:422: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'logfoldchanges'] = np.log2(
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'names'] = self.var_names[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:398: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'scores'] = scores[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:401: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals'] = pvals[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:411: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]
/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:422: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  self.stats[group_name, 'logfoldchanges'] = np.log2(</code></pre>
</div>
</div>
<div id="8a829806-5df5-457f-bbfc-6ae7e459f172" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.uns[<span class="st">"dea_leiden"</span>].keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>dict_keys(['params', 'names', 'scores', 'pvals', 'pvals_adj', 'logfoldchanges'])</code></pre>
</div>
</div>
<div id="c6ef562b-f629-454f-8e8a-925a1b7be1b2" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="op">=</span> pd.DataFrame(adata_raw_norm.uns[<span class="st">"dea_leiden"</span>][<span class="st">"names"</span>])</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>marker_genes.columns <span class="op">=</span> [<span class="st">"Cluster"</span> <span class="op">+</span> <span class="bu">str</span>(x) <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(adata_raw_norm.obs[<span class="st">"leiden"</span>].unique()))]</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>marker_genes.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="32">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Cluster0</th>
<th data-quarto-table-cell-role="th">Cluster1</th>
<th data-quarto-table-cell-role="th">Cluster2</th>
<th data-quarto-table-cell-role="th">Cluster3</th>
<th data-quarto-table-cell-role="th">Cluster4</th>
<th data-quarto-table-cell-role="th">Cluster5</th>
<th data-quarto-table-cell-role="th">Cluster6</th>
<th data-quarto-table-cell-role="th">Cluster7</th>
<th data-quarto-table-cell-role="th">Cluster8</th>
<th data-quarto-table-cell-role="th">Cluster9</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Cluster14</th>
<th data-quarto-table-cell-role="th">Cluster15</th>
<th data-quarto-table-cell-role="th">Cluster16</th>
<th data-quarto-table-cell-role="th">Cluster17</th>
<th data-quarto-table-cell-role="th">Cluster18</th>
<th data-quarto-table-cell-role="th">Cluster19</th>
<th data-quarto-table-cell-role="th">Cluster20</th>
<th data-quarto-table-cell-role="th">Cluster21</th>
<th data-quarto-table-cell-role="th">Cluster22</th>
<th data-quarto-table-cell-role="th">Cluster23</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>RPL21</td>
<td>HBA2</td>
<td>S100A8</td>
<td>SLC25A37</td>
<td>CD74</td>
<td>HBB</td>
<td>FTH1</td>
<td>DNTT</td>
<td>RPS4X</td>
<td>CD74</td>
<td>...</td>
<td>RPS27</td>
<td>CD52</td>
<td>CD79B</td>
<td>NKG7</td>
<td>PRDX2</td>
<td>ATPIF1</td>
<td>MPO</td>
<td>HBD</td>
<td>SLC25A37</td>
<td>SSR4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>RPS27</td>
<td>HBB</td>
<td>S100A9</td>
<td>HBD</td>
<td>CST3</td>
<td>HBA2</td>
<td>FTL</td>
<td>IGLL1</td>
<td>RPS18</td>
<td>MS4A1</td>
<td>...</td>
<td>RPL21</td>
<td>TMSB4X</td>
<td>IGHM</td>
<td>B2M</td>
<td>GYPB</td>
<td>APOC1</td>
<td>AZU1</td>
<td>GYPA</td>
<td>HBD</td>
<td>MZB1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>RPS29</td>
<td>HBA1</td>
<td>LYZ</td>
<td>HMBS</td>
<td>HLA-DRA</td>
<td>HBA1</td>
<td>TYROBP</td>
<td>VPREB1</td>
<td>RPS8</td>
<td>CD79A</td>
<td>...</td>
<td>RPS29</td>
<td>TRAC</td>
<td>TCL1A</td>
<td>HLA-B</td>
<td>HEMGN</td>
<td>NME4</td>
<td>SRGN</td>
<td>HMBS</td>
<td>SLC4A1</td>
<td>HSP90B1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>RPL13</td>
<td>SNCA</td>
<td>S100A6</td>
<td>SLC4A1</td>
<td>HLA-DPB1</td>
<td>BPGM</td>
<td>CTSS</td>
<td>PTMA</td>
<td>RPLP0</td>
<td>HLA-DRA</td>
<td>...</td>
<td>RPL13</td>
<td>LTB</td>
<td>CD24</td>
<td>CCL5</td>
<td>AHSP</td>
<td>TMEM14C</td>
<td>ELANE</td>
<td>SLC25A37</td>
<td>BNIP3L</td>
<td>FKBP11</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>RPLP2</td>
<td>BNIP3L</td>
<td>S100A4</td>
<td>SLC2A1</td>
<td>HLA-DRB1</td>
<td>UBB</td>
<td>FCER1G</td>
<td>HMGB1</td>
<td>RPS24</td>
<td>CD37</td>
<td>...</td>
<td>RPLP2</td>
<td>B2M</td>
<td>RCSD1</td>
<td>HLA-C</td>
<td>CA2</td>
<td>FAM178B</td>
<td>CFD</td>
<td>SLC4A1</td>
<td>AHSP</td>
<td>SEC11C</td>
</tr>
</tbody>
</table>

<p>5 rows × 24 columns</p>
</div>
</div>
</div>
<div id="42e574b8-b97b-4b75-b92b-ffa674c206d7" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>marker_genes_pvals <span class="op">=</span> pd.DataFrame(adata_raw_norm.uns[<span class="st">"dea_leiden"</span>][<span class="st">"logfoldchanges"</span>])</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>marker_genes_pvals.columns <span class="op">=</span> [<span class="st">"Cluster"</span> <span class="op">+</span> <span class="bu">str</span>(x) <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(adata_raw_norm.obs[<span class="st">"leiden"</span>].unique()))]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>marker_genes_pvals.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="34">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Cluster0</th>
<th data-quarto-table-cell-role="th">Cluster1</th>
<th data-quarto-table-cell-role="th">Cluster2</th>
<th data-quarto-table-cell-role="th">Cluster3</th>
<th data-quarto-table-cell-role="th">Cluster4</th>
<th data-quarto-table-cell-role="th">Cluster5</th>
<th data-quarto-table-cell-role="th">Cluster6</th>
<th data-quarto-table-cell-role="th">Cluster7</th>
<th data-quarto-table-cell-role="th">Cluster8</th>
<th data-quarto-table-cell-role="th">Cluster9</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Cluster14</th>
<th data-quarto-table-cell-role="th">Cluster15</th>
<th data-quarto-table-cell-role="th">Cluster16</th>
<th data-quarto-table-cell-role="th">Cluster17</th>
<th data-quarto-table-cell-role="th">Cluster18</th>
<th data-quarto-table-cell-role="th">Cluster19</th>
<th data-quarto-table-cell-role="th">Cluster20</th>
<th data-quarto-table-cell-role="th">Cluster21</th>
<th data-quarto-table-cell-role="th">Cluster22</th>
<th data-quarto-table-cell-role="th">Cluster23</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>1.696161</td>
<td>4.536685</td>
<td>6.127852</td>
<td>3.926013</td>
<td>3.412972</td>
<td>4.632360</td>
<td>2.845425</td>
<td>4.403666</td>
<td>1.796766</td>
<td>3.350841</td>
<td>...</td>
<td>1.718245</td>
<td>2.538189</td>
<td>4.155026</td>
<td>5.839576</td>
<td>3.842825</td>
<td>2.919723</td>
<td>7.224152</td>
<td>4.162687</td>
<td>3.580635</td>
<td>3.880045</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1.909575</td>
<td>4.749588</td>
<td>5.971000</td>
<td>4.183972</td>
<td>3.933990</td>
<td>4.355470</td>
<td>3.283564</td>
<td>4.142626</td>
<td>1.714990</td>
<td>5.916070</td>
<td>...</td>
<td>1.482474</td>
<td>1.988572</td>
<td>3.510608</td>
<td>1.826526</td>
<td>3.914904</td>
<td>5.641148</td>
<td>6.555845</td>
<td>3.788034</td>
<td>3.828166</td>
<td>4.938936</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>1.893222</td>
<td>4.461418</td>
<td>6.040055</td>
<td>3.383904</td>
<td>3.215807</td>
<td>4.279613</td>
<td>4.806640</td>
<td>3.918157</td>
<td>1.856143</td>
<td>3.856486</td>
<td>...</td>
<td>1.645742</td>
<td>3.124133</td>
<td>4.694790</td>
<td>2.050864</td>
<td>3.797841</td>
<td>3.448345</td>
<td>4.227748</td>
<td>3.277330</td>
<td>2.918315</td>
<td>3.337664</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>1.828848</td>
<td>2.532328</td>
<td>3.790318</td>
<td>3.315197</td>
<td>3.314304</td>
<td>2.452660</td>
<td>4.474485</td>
<td>2.222489</td>
<td>1.779858</td>
<td>2.956043</td>
<td>...</td>
<td>1.580002</td>
<td>2.921369</td>
<td>4.179593</td>
<td>6.373088</td>
<td>3.689547</td>
<td>3.167665</td>
<td>6.572220</td>
<td>3.602345</td>
<td>2.388110</td>
<td>4.894440</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1.628970</td>
<td>2.246099</td>
<td>3.793867</td>
<td>3.430458</td>
<td>3.418615</td>
<td>1.447062</td>
<td>4.434198</td>
<td>2.761628</td>
<td>1.619545</td>
<td>2.928020</td>
<td>...</td>
<td>1.443913</td>
<td>1.548262</td>
<td>3.441401</td>
<td>2.135884</td>
<td>4.070135</td>
<td>6.136931</td>
<td>3.634571</td>
<td>3.219230</td>
<td>3.156719</td>
<td>4.655691</td>
</tr>
</tbody>
</table>

<p>5 rows × 24 columns</p>
</div>
</div>
</div>
<div id="4d17699b-3ca2-4d35-8c63-d399db0b089d" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">50</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_dotplot(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    adata_raw_norm, groupby<span class="op">=</span><span class="st">"leiden"</span>, standard_scale<span class="op">=</span><span class="st">"var"</span>, n_genes<span class="op">=</span><span class="dv">5</span>, key<span class="op">=</span><span class="st">"dea_leiden"</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: dendrogram data not found (using key=dendrogram_leiden). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/plotting/_dotplot.py:747: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap', 'norm' will be ignored
  dot_ax.scatter(x, y, **kwds)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-11-output-3.png" width="1853" height="413" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As you can see above, a lot of the differentially expressed genes are highly expressed in multiple clusters. We can filter the differentially expressed genes to select for more cluster-specific differentially expressed genes:</p>
<div id="0533555a-c030-4f21-bb9f-bdae3d4c7e11" class="cell" data-executioninfo="{&quot;elapsed&quot;:37816,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340534478,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="6934f254-1a02-419f-962d-82c59744337a" data-execution_count="8">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sc.tl.filter_rank_genes_groups(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    adata_raw_norm,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    min_in_group_fraction<span class="op">=</span><span class="fl">0.2</span>,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    max_out_group_fraction<span class="op">=</span><span class="fl">0.2</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    key<span class="op">=</span><span class="st">"dea_leiden"</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    key_added<span class="op">=</span><span class="st">"dea_leiden_filtered"</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="688ef6f1-311f-44d9-95ae-8b1583d83afa" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_dotplot(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    adata_raw_norm, groupby<span class="op">=</span><span class="st">"leiden"</span>, standard_scale<span class="op">=</span><span class="st">"var"</span>, n_genes<span class="op">=</span><span class="dv">5</span>, key<span class="op">=</span><span class="st">"dea_leiden_filtered"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/plotting/_dotplot.py:747: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap', 'norm' will be ignored
  dot_ax.scatter(x, y, **kwds)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-13-output-2.png" width="1853" height="416" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Exercise 1</strong>: Visualize marker genes on the UMAP or tSNE representation. Try to find 3-4 marker genes that are indeed specific to a particular cluster. Are there any clusters that do not seem to have unique marker genes? Are there any clusters containing markers that are only specific to a portion of the cluster? Marker genes should uniformly define cells “everywhere” in a cluster in UMAP space, otherwise the cluster might actually be two!</p>
<div id="4804f79d-960f-4431-87d6-ae6920130d24" class="cell" data-executioninfo="{&quot;elapsed&quot;:1510,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340543187,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="3650a62f-a65a-46d5-d377-86ef2246d66a" data-execution_count="10">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"CD74"</span>, <span class="st">"SSR4"</span>, <span class="st">"CA2"</span>, <span class="st">"HBA2"</span>, <span class="st">"CST3"</span>, <span class="st">"CD37"</span>, <span class="st">"IL32"</span>, <span class="st">"leiden_res0_5"</span>],</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    vmax<span class="op">=</span><span class="st">"p99"</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    legend_loc<span class="op">=</span><span class="st">"on data"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    frameon<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    cmap<span class="op">=</span><span class="st">"coolwarm"</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/alex/anaconda3/envs/sctp/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-14-output-2.png" width="791" height="353" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Exercise 2</strong>: Let’s take a few steps back to understand all of the previous steps a little bit better! The number of genes selected by the highly_variable_genes function can significantly impact your ability to cluster. Too few genes and you cannot discriminate between different cell types, too many genes and you capture lots of noisy clusters! Try repeating the previous analysis with either 200 or 5000 highly variable genes, naming the AnnData object differently (i.e.&nbsp;adata_200genes) to avoid overwriting your previous results. Transfer the metadata for the new cluster labels to the original AnnData object’s metadata at adata.obs and compare on the UMAP. Are the clusters different?</p>
<p>Once you have settled on the parameters for the dimensionality reduction and clustering steps, it is time to begin annotating your clusters with cell types. This is normally a challenging step! When you are not too familiar with the marker genes for a particular cluster, a good starting point is simply to Google a strong marker gene and understand its function. Other tools that might be useful include EnrichR and GSEAPy. - https://maayanlab.cloud/Enrichr/ - https://gseapy.readthedocs.io/en/latest/gseapy_example.html#2.-Enrichr-Example</p>
<p>Fortunately in our case, we will try automated cell type annotations!</p>
<section id="automated-cell-type-annotation" class="level1">
<h1>Automated cell type annotation</h1>
<p><strong>Exercise 3:</strong> The methods discussed here focus on automated data annotation, distinct from manual methods. Unlike the previously detailed approach, these methods automate data annotation. They operate on different principles, using predefined markers or trained on comprehensive scRNA-seq datasets. It’s vital to note that automated annotations can vary in quality. Thus, they should be seen as a starting point rather than a final solution. <a href="https://www.csbj.org/article/S2001-03702100019-2/fulltext">Pasquini et al., 2021</a> and <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1795-z">Abdelaal et al., 2019</a> offer extensive discussions on automated annotation methods.</p>
<p>Quality depends on:</p>
<ul>
<li><p>Classifier Choice: Various classifier types perform similarly, with neural networks <strong>not</strong> necessarily outperforming linear models [<a href="https://www.csbj.org/article/S2001-03702100019-2/fulltext">1</a>, <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1795-z">2</a>, <a href="https://academic.oup.com/bib/article/22/5/bbab035/6145135?login=false">3</a>].</p></li>
<li><p>Training Data Quality: Annotation quality relies on the quality of the training data. Poorly annotated or noisy training data can impact the classifier.</p></li>
<li><p>Data Similarity: Similarity between your data and the classifier’s training data matters. Cross-dataset models often provide better annotations. For example, CellTypist, trained on diverse lung datasets, is likely to perform well on new lung data.</p></li>
</ul>
<p>While classifiers have limitations, they offer advantages like rapid annotation, leveraging previous studies, and promoting standardized terminology. Ensuring robust uncertainty measures to quantify annotation reliability is crucial.</p>
<p>Many classification methods rely on a limited set of genes, typically just 1 to ~10 marker genes per cell type. An alternative approach utilizes classifiers that consider a more extensive gene set, often several thousands or more. These classifiers are trained on previously annotated datasets or atlases. Notable examples include CellTypist <a href="https://www.science.org/doi/full/10.1126/science.abl5197">Conde et al., 2022</a> and Clustifyr <a href="https://f1000research.com/articles/9-223/v2">Fu et al., 2020</a>.</p>
<p>Let’s explore CellTypist for our data. Referring to the <a href="https://www.celltypist.org/tutorials">CellTypist tutorial</a>, we should prepare our data by normalizing counts to 10,000 counts per cell and subsequently applying a log1p transformation. So we need to re-normalize our data, without our logarithm shift approach, but with a more classical ‘Counts per ten-thousand’.</p>
<div id="b403385d-e740-43a2-a00a-421566146129" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> celltypist</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> celltypist <span class="im">import</span> models</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5eb51d45-e901-427a-9712-26fb335065c9" class="cell" data-executioninfo="{&quot;elapsed&quot;:1690,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340621373,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="d2d313e4-b28f-444d-e69b-30c3db855303" data-execution_count="12">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>adata_celltypist <span class="op">=</span> adata_raw.copy()  <span class="co"># make a copy of our adata</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    adata_celltypist, counts_per_cell_after<span class="op">=</span><span class="fl">10000.0</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>)  <span class="co"># normalize to 10,000 counts per cell</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata_celltypist)  <span class="co"># log-transform</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make .X dense instead of sparse, for compatibility with celltypist:</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>adata_celltypist.X <span class="op">=</span> adata_celltypist.X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we will load the model directly from our folder on google drive, where we can find the model trained. Alternatively, CellTypist method propose a panel of models that can be download directly from python using <code>models.download_models(force_update = True)</code>. The idea is of course to use a model that match our biological context, and for pre-trained model-based method like CellTypist, it is possible that your biological context is not available. In that situation, there is no other options than opting for manual annotations.</p>
<p>There are two models that might be relevant for this particular dataset we are working with. Let’s download both of them and try each one for the classification.</p>
<div id="e54273de-dcbf-46f6-aa44-c5e960c404cc" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>models.download_models(</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    force_update<span class="op">=</span><span class="va">True</span>, model<span class="op">=</span>[<span class="st">"Immune_All_Low.pkl"</span>, <span class="st">"Immune_All_High.pkl"</span>]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>📜 Retrieving model list from server https://celltypist.cog.sanger.ac.uk/models/models.json
📚 Total models in list: 48
📂 Storing models in /home/alex/.celltypist/data/models
💾 Total models to download: 2
💾 Downloading model [1/2]: Immune_All_Low.pkl
💾 Downloading model [2/2]: Immune_All_High.pkl</code></pre>
</div>
</div>
<div id="32e0eb3f-3fac-46b3-969f-8a379733cb20" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>model_low <span class="op">=</span> models.Model.load(model<span class="op">=</span><span class="st">"Immune_All_Low.pkl"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>model_high <span class="op">=</span> models.Model.load(model<span class="op">=</span><span class="st">"Immune_All_High.pkl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For each of these, we can see which cell types it includes to see if bone marrow cell types are included:</p>
<div id="2d9413f2-bc17-4c3d-9b44-0d3307178ddf" class="cell" data-executioninfo="{&quot;elapsed&quot;:2,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340621804,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="78732559-38c4-4a2a-b288-6de932a266a3" data-execution_count="15">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can print all the cell types covererd by the model</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>model_low.cell_types</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<pre><code>array(['Age-associated B cells', 'Alveolar macrophages', 'B cells',
       'CD16+ NK cells', 'CD16- NK cells', 'CD8a/a', 'CD8a/b(entry)',
       'CMP', 'CRTAM+ gamma-delta T cells', 'Classical monocytes',
       'Cycling B cells', 'Cycling DCs', 'Cycling NK cells',
       'Cycling T cells', 'Cycling gamma-delta T cells',
       'Cycling monocytes', 'DC', 'DC precursor', 'DC1', 'DC2', 'DC3',
       'Double-negative thymocytes', 'Double-positive thymocytes', 'ELP',
       'ETP', 'Early MK', 'Early erythroid', 'Early lymphoid/T lymphoid',
       'Endothelial cells', 'Epithelial cells', 'Erythrocytes',
       'Erythrophagocytic macrophages', 'Fibroblasts',
       'Follicular B cells', 'Follicular helper T cells', 'GMP',
       'Germinal center B cells', 'Granulocytes', 'HSC/MPP',
       'Hofbauer cells', 'ILC', 'ILC precursor', 'ILC1', 'ILC2', 'ILC3',
       'Intermediate macrophages', 'Intestinal macrophages',
       'Kidney-resident macrophages', 'Kupffer cells',
       'Large pre-B cells', 'Late erythroid', 'MAIT cells', 'MEMP', 'MNP',
       'Macrophages', 'Mast cells', 'Megakaryocyte precursor',
       'Megakaryocyte-erythroid-mast cell progenitor',
       'Megakaryocytes/platelets', 'Memory B cells',
       'Memory CD4+ cytotoxic T cells', 'Mid erythroid', 'Migratory DCs',
       'Mono-mac', 'Monocyte precursor', 'Monocytes', 'Myelocytes',
       'NK cells', 'NKT cells', 'Naive B cells',
       'Neutrophil-myeloid progenitor', 'Neutrophils',
       'Non-classical monocytes', 'Plasma cells', 'Plasmablasts',
       'Pre-pro-B cells', 'Pro-B cells',
       'Proliferative germinal center B cells', 'Promyelocytes',
       'Regulatory T cells', 'Small pre-B cells', 'T(agonist)',
       'Tcm/Naive cytotoxic T cells', 'Tcm/Naive helper T cells',
       'Tem/Effector helper T cells', 'Tem/Effector helper T cells PD1+',
       'Tem/Temra cytotoxic T cells', 'Tem/Trm cytotoxic T cells',
       'Transitional B cells', 'Transitional DC', 'Transitional NK',
       'Treg(diff)', 'Trm cytotoxic T cells', 'Type 1 helper T cells',
       'Type 17 helper T cells', 'gamma-delta T cells', 'pDC',
       'pDC precursor'], dtype=object)</code></pre>
</div>
</div>
<div id="c2454bc9-e39c-4dfb-88a1-6769a7337562" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can print all the cell types covererd by the model</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>model_high.cell_types</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>array(['B cells', 'B-cell lineage', 'Cycling cells', 'DC', 'DC precursor',
       'Double-negative thymocytes', 'Double-positive thymocytes', 'ETP',
       'Early MK', 'Endothelial cells', 'Epithelial cells',
       'Erythrocytes', 'Erythroid', 'Fibroblasts', 'Granulocytes',
       'HSC/MPP', 'ILC', 'ILC precursor', 'MNP', 'Macrophages',
       'Mast cells', 'Megakaryocyte precursor',
       'Megakaryocytes/platelets', 'Mono-mac', 'Monocyte precursor',
       'Monocytes', 'Myelocytes', 'Plasma cells', 'Promyelocytes',
       'T cells', 'pDC', 'pDC precursor'], dtype=object)</code></pre>
</div>
</div>
<p>The model_high seems to have fewer cell types, let’s start with that for obtaining broader cell type categories.</p>
<p><strong>Exercise 4:</strong> Use the <code>celltypist.annotate</code> function to predict cell types using the model <code>model_high</code> and <code>marjority_voting=True</code>: Save the result to a variable called <code>predictions_high</code>.</p>
<button onclick="toggleVisibility('answer4')">
Click for Answer
</button>
<div id="answer4" style="display:none;">
<strong>Answer:</strong>
<pre>    predictions_high = celltypist.annotate(
    adata_celltypist, model=model_high, majority_voting=True)

    🔬 Input data has 5459 cells and 10839 genes
    🔗 Matching reference genes in the model
    🧬 3729 features used for prediction
    ⚖️ Scaling input data
    🖋️ Predicting labels
    ✅ Prediction done!
    👀 Can not detect a neighborhood graph, will construct one before the over-clustering
    ⛓️ Over-clustering input data with resolution set to 10
    🗳️ Majority voting the predictions
    ✅ Majority voting done!
</pre>
<script>
function toggleVisibility(id) {
   var element = document.getElementById(id);
   if (element.style.display === 'none') {
       element.style.display = 'block';
   } else {
       element.style.display = 'none';
   }
}
</script>
</div>
<div id="55536379-cd60-489b-ae6f-3c7f576d7a09" class="cell" data-executioninfo="{&quot;elapsed&quot;:19,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340636224,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="8241cd8f-f8b2-4e64-8d30-2fd265f74494" data-execution_count="18">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>predictions_high_adata <span class="op">=</span> predictions_high.to_adata()</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>predictions_high_adata.obs[[<span class="st">'majority_voting'</span>, <span class="st">'conf_score'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">majority_voting</th>
<th data-quarto-table-cell-role="th">conf_score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">TTCCCAGCAGACAAAT-1</td>
<td>B-cell lineage</td>
<td>0.990278</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GTGTGCGGTGTTTGGT-1</td>
<td>B-cell lineage</td>
<td>0.986694</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GATCAGTTCTTTAGTC-1</td>
<td>Erythroid</td>
<td>0.984309</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GTCTTCGAGAAGATTC-1</td>
<td>Monocytes</td>
<td>0.999283</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TGCTACCTCGTTACAG-1</td>
<td>pDC</td>
<td>0.996167</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TCATTACCAAGCGTAG-1</td>
<td>Monocytes</td>
<td>0.938815</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">CAGATCAAGATGCGAC-1</td>
<td>Monocytes</td>
<td>0.717744</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">CGTCAGGAGTGCTGCC-1</td>
<td>Erythroid</td>
<td>0.992524</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ACACTGAGTTGATTCG-1</td>
<td>Erythroid</td>
<td>0.933983</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GAAGCAGTCGAACTGT-1</td>
<td>T cells</td>
<td>0.997989</td>
</tr>
</tbody>
</table>

<p>5459 rows × 2 columns</p>
</div>
</div>
</div>
<div id="4873ac7a-34ba-4fe6-823d-428ab02871c8" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obs[<span class="st">"celltypist_annotations_high"</span>] <span class="op">=</span> predictions_high_adata.obs[<span class="st">"majority_voting"</span>]</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obs[<span class="st">"celltypist_conf_score_high"</span>] <span class="op">=</span> predictions_high_adata.obs[<span class="st">"conf_score"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Exercise 5:</strong> Now do the same (<code>celltypist.annotate</code>) for the finer-grained annotations. Save the result to a variable called <code>predictions_low</code>.</p>
<button onclick="toggleVisibility('answer5')">
Click for Answer
</button>
<div id="answer5" style="display:none;">
<strong>Answer:</strong>
<pre>    predictions_low = celltypist.annotate(
    adata_celltypist, model=model_low, majority_voting=True)

    🔬 Input data has 5459 cells and 10839 genes
    🔗 Matching reference genes in the model
    🧬 3729 features used for prediction
    ⚖️ Scaling input data
    🖋️ Predicting labels
    ✅ Prediction done!
    👀 Can not detect a neighborhood graph, will construct one before the over-clustering
    ⛓️ Over-clustering input data with resolution set to 10
    🗳️ Majority voting the predictions
    ✅ Majority voting done!
</pre>
<script>
function toggleVisibility(id) {
   var element = document.getElementById(id);
   if (element.style.display === 'none') {
       element.style.display = 'block';
   } else {
       element.style.display = 'none';
   }
}
</script>
</div>
<div id="a1f1e332-9a71-4766-b28f-05ef0fff96b5" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>predictions_low_adata <span class="op">=</span> predictions_low.to_adata()</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>predictions_low_adata.obs[[<span class="st">'majority_voting'</span>, <span class="st">'conf_score'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">majority_voting</th>
<th data-quarto-table-cell-role="th">conf_score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">TTCCCAGCAGACAAAT-1</td>
<td>Large pre-B cells</td>
<td>0.501054</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GTGTGCGGTGTTTGGT-1</td>
<td>Large pre-B cells</td>
<td>0.827030</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GATCAGTTCTTTAGTC-1</td>
<td>Late erythroid</td>
<td>0.897641</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GTCTTCGAGAAGATTC-1</td>
<td>Monocytes</td>
<td>0.982690</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TGCTACCTCGTTACAG-1</td>
<td>pDC</td>
<td>0.996716</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TCATTACCAAGCGTAG-1</td>
<td>Classical monocytes</td>
<td>0.871458</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">CAGATCAAGATGCGAC-1</td>
<td>Classical monocytes</td>
<td>0.318097</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">CGTCAGGAGTGCTGCC-1</td>
<td>Mid erythroid</td>
<td>0.361174</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ACACTGAGTTGATTCG-1</td>
<td>Late erythroid</td>
<td>0.935759</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GAAGCAGTCGAACTGT-1</td>
<td>Tcm/Naive helper T cells</td>
<td>0.155408</td>
</tr>
</tbody>
</table>

<p>5459 rows × 2 columns</p>
</div>
</div>
</div>
<p>And we save our predictions to our AnnData object:</p>
<div id="0d4ef87e-92d1-4b6b-8a20-0ac344c09981" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obs[<span class="st">"celltypist_annotations_low"</span>] <span class="op">=</span> predictions_low_adata.obs[<span class="st">"majority_voting"</span>]</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>adata_raw_norm.obs[<span class="st">"celltypist_conf_score_low"</span>] <span class="op">=</span> predictions_low_adata.obs[<span class="st">"conf_score"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>CellTypist annotations can then be visualized on the UMAP embedding:</p>
<div id="8f9d846b-9f1d-4a07-825e-09922afb59b4" class="cell" data-executioninfo="{&quot;elapsed&quot;:1596,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340637804,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="1eaca042-bffb-48a5-807e-dc6f865a9d44" data-execution_count="23">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>, facecolor<span class="op">=</span><span class="st">'white'</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    adata_raw_norm,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"celltypist_annotations_low"</span>, <span class="st">"celltypist_annotations_high"</span>],</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    frameon<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    sort_order<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    wspace<span class="op">=</span><span class="fl">1.2</span>,</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-25-output-1.png" width="1240" height="286" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Also, each cell gets a prediction score:</p>
<div id="0dfacb9f-c9d6-4934-951b-149b36ad8163" class="cell" data-executioninfo="{&quot;elapsed&quot;:684,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340638480,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="6f3c9170-a120-492b-b43b-5dc3adb098d3" data-execution_count="24">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>    adata_raw_norm,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"celltypist_conf_score_low"</span>, <span class="st">"celltypist_conf_score_high"</span>],</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    frameon<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    sort_order<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>    wspace<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-26-output-1.png" width="1047" height="280" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>One way of getting a feeling for the quality of these annotations is by looking if the observed cell type similarities correspond to our expectations:</p>
<div id="8160941b-a38b-4fd1-b7f4-689651cc1b50" class="cell" data-executioninfo="{&quot;elapsed&quot;:7,&quot;status&quot;:&quot;ok&quot;,&quot;timestamp&quot;:1694340638480,&quot;user&quot;:{&quot;displayName&quot;:&quot;Alex Lederer&quot;,&quot;userId&quot;:&quot;09653581722857927737&quot;},&quot;user_tz&quot;:-120}" data-outputid="a46856f4-1aab-41a1-dd0f-29b2fcbf9bcc" data-execution_count="25">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sc.pl.dendrogram(adata_raw_norm, groupby<span class="op">=</span><span class="st">"celltypist_annotations_low"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: dendrogram data not found (using key=dendrogram_celltypist_annotations_low). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-27-output-2.png" width="265" height="462" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><strong>Exercise 6:</strong> Can you identify marker genes for each annotated cell type from <code>celltypist_annotations_high</code>? Do this similarly to how you did for the leiden clusters above, i.e., with the <code>sc.tl.rank_genes_groups</code> function, grouped by the annotated cell type. Can you convert the results into a pandas dataframe?</p>
<button onclick="toggleVisibility('answer6')">
Click for Answer
</button>
<div id="answer6" style="display:none;">
<strong>Answer:</strong> You should write the following code:
<pre>    sc.tl.rank_genes_groups(adata_raw_norm, use_raw=False, 
          groupby="celltypist_annotations_high", method="wilcoxon", key_added="dea_celltype")

    marker_genes = pd.DataFrame(adata_raw_norm.uns["dea_celltype"]["names"])
    marker_genes.columns = sorted(list(adata_raw_norm.obs["celltypist_annotations_high"].unique()))
    marker_genes.head()
</pre>
<script>
function toggleVisibility(id) {
   var element = document.getElementById(id);
   if (element.style.display === 'none') {
       element.style.display = 'block';
   } else {
       element.style.display = 'none';
   }
}
</script>
</div>
</section>
<section id="another-way-to-annotate-with-label-transfer-from-a-reference-dataset" class="level1">
<h1>Another way to annotate: with label transfer from a reference dataset!</h1>
<p>In addition to automated label transfer methods using machine learning, it is possible to transfer cell type labels acquired (either automatically or manually) from a reference dataset to a new, unannotated dataset, using the <code>sc.tl.ingest</code> function.</p>
<p>The <a href="https://scanpy.readthedocs.io/en/latest/api/scanpy.tl.ingest.html">ingest</a> function assumes an annotated reference dataset that captures the biological variability of interest. The rational is to fit a model on the reference data and use it to project new data. For the time being, this model is a PCA combined with a neighbor lookup search tree, for which we use UMAP’s implementation <a href="https://arxiv.org/abs/1802.03426">[McInnes18]</a>. Similar PCA-based integrations have been used before, for instance, in <a href="https://doi.org/10.1101/467886">[Weinreb18]</a>.</p>
<ul>
<li>As <code>ingest</code> is simple and the procedure clear, the workflow is transparent and fast.</li>
<li>The function leaves the data matrix itself invariant, unlike many integration methods.</li>
<li>The function also solves the label mapping problem and maintains an embedding that might have desired properties like specific clusters or trajectories.</li>
</ul>
<p>We refer to this <em>asymmetric</em> dataset integration as <em>ingesting</em> annotations from an annotated reference <code>adata_ref</code> into an <code>adata</code> that still lacks this annotation. It is different from learning a joint representation that integrates datasets in a symmetric way as in CCA (e.g.&nbsp;from Seurat).</p>
<p>Take a look at tools in the <a href="https://scanpy.readthedocs.io/en/latest/external/#data-integration">external API</a> or at the <a href="https://scanpy.readthedocs.io/en/latest/ecosystem/#data-integration">ecoystem page</a> for scanpy to read about other related data integration and label transfer tools.</p>
<p>Let’s evaluate the role of the <code>ingest</code> function by considering the case where we only use celltypist to annotate cells from one sample: PBMMC_1. We can use <code>ingest</code> to then transfer the labels to another sample, PBMMC_3.</p>
<div id="ec7e9d3c-f148-4197-b9b3-4170510f2e68" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reference dataset</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>sample1_adata <span class="op">=</span> adata_raw_norm[adata_raw_norm.obs[<span class="st">"sample"</span>]<span class="op">==</span><span class="st">"PBMMC_1"</span>].copy()</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Unlabeled dataset to transfer cell type labels to</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>sample3_adata <span class="op">=</span> adata_raw_norm[adata_raw_norm.obs[<span class="st">"sample"</span>]<span class="op">==</span><span class="st">"PBMMC_3"</span>].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In order for <code>ingest</code> to transfer the labels, we need to first remove the labels obtained with celltypist from our <code>sample3_adata</code> object. In the meantime, we can store them in a variable <code>old_annotations_sample3</code>:</p>
<div id="636b0db3-09fc-4907-9937-368569959cb8" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>old_annotations_sample3 <span class="op">=</span> np.array(sample3_adata.obs[<span class="st">"celltypist_annotations_high"</span>])</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> sample3_adata.obs[<span class="st">"celltypist_annotations_high"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We need to quickly repeat the analysis steps we have previously performed before, but this time on the two samples (PBMMC_1; i.e.&nbsp;<code>sample1_adata</code> and PBMMC_3; i.e.&nbsp;<code>sample3_adata</code>) independently:</p>
<p><strong>Exercise 7:</strong> Repeat the following steps, once with <code>sample1_adata</code> and once with <code>sample3_adata</code>:</p>
<p><code>sc.pp.highly_variable_genes</code> with <code>subset=True</code> and <code>n_top_genes=3000</code></p>
<p><code>sc.pp.pca</code></p>
<p><code>sc.pp.neighbors</code></p>
<p><code>sc.tl.umap</code></p>
<button onclick="toggleVisibility('answer7')">
Click for Answer
</button>
<div id="answer7" style="display:none;">
<strong>Answer:</strong> You should write the following code:
<pre>    sc.pp.highly_variable_genes(sample1_adata, subset=True, n_top_genes=3000)
    sc.pp.pca(sample1_adata)
    sc.pp.neighbors(sample1_adata)
    sc.tl.umap(sample1_adata)


    sc.pp.highly_variable_genes(sample3_adata, subset=True, n_top_genes=3000)
    sc.pp.pca(sample3_adata)
    sc.pp.neighbors(sample3_adata)
    sc.tl.umap(sample3_adata)
</pre>
<script>
function toggleVisibility(id) {
   var element = document.getElementById(id);
   if (element.style.display === 'none') {
       element.style.display = 'block';
   } else {
       element.style.display = 'none';
   }
}
</script>
</div>
<p>Visualize your generated UMAPs for each individual sample:</p>
<div id="b66b0e31-633f-4d3e-b92f-e55046771451" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(sample1_adata, color<span class="op">=</span><span class="st">"celltypist_annotations_high"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-30-output-1.png" width="421" height="296" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="bbf6ba87-f57f-4526-9884-11284ede855f" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(sample3_adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-31-output-1.png" width="284" height="278" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>To use <code>sc.tl.ingest</code>, the datasets need to be defined on the same variables.</p>
<div id="8b55b5ad-1a00-4f67-b3df-9ff1e378b3a3" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>var_names <span class="op">=</span> sample1_adata.var_names.intersection(sample3_adata.var_names)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>sample1_adata <span class="op">=</span> sample1_adata[:, var_names]</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>sample3_adata <span class="op">=</span> sample3_adata[:, var_names]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can finally map labels and embeddings from <code>sample1_adata</code> (our reference) to <code>sample3_adata</code> (our unannotated new dataset) based on a chosen representation.</p>
<p><strong>Exercise 8:</strong> Run the <code>sc.tl.ingest</code> command, with <code>sample3_adata</code> as your unlabeled dataset and <code>sample1_adata</code> as your reference dataset. Transfer the label <code>celltypist_annotations_high</code> using the <code>pca</code> as the embedding method.</p>
<button onclick="toggleVisibility('answer8')">
Click for Answer
</button>
<div id="answer8" style="display:none;">
<strong>Answer:</strong> You should write the following code:
<pre>    sc.tl.ingest(adata=sample3_adata, 
                 adata_ref=sample1_adata, 
                 obs="celltypist_annotations_high",
                 embedding_method='pca')
</pre>
<script>
function toggleVisibility(id) {
   var element = document.getElementById(id);
   if (element.style.display === 'none') {
       element.style.display = 'block';
   } else {
       element.style.display = 'none';
   }
}
</script>
</div>
<p>Finally, we want to rename the labels obtained for <code>sample3_adata</code> from <code>celltypist</code> and <code>ingest</code>, and compare them to each other on the UMAP space:</p>
<div id="ce241c06-f52c-4e4e-93c0-ca43ba146cab" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sample3_adata.obs[<span class="st">"celltypist_labels"</span>] <span class="op">=</span> old_annotations_sample3</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>sample3_adata.obs[<span class="st">"ingest_labels"</span>] <span class="op">=</span> sample3_adata.obs[<span class="st">"celltypist_annotations_high"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="60b73b7a-22d5-4d7f-bdc6-824fe24c39ed" class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(sample3_adata, color<span class="op">=</span>[<span class="st">'celltypist_labels'</span>, <span class="st">'ingest_labels'</span>], wspace<span class="op">=</span><span class="fl">0.50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-4_visualization_files/figure-html/cell-34-output-1.png" width="907" height="299" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>By comparing the <code>ingest_labels</code> annotation with <code>celltypist_labels</code>, we see that the data has been reasonably mapped.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/sib-swiss\.github\.io\/single-cell-python-training\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>